File: Practica 16/Quicksort.cpp
#include<iostream>
#include<stdlib.h>
#include<string.h>
#include <ctime>
#include<fstream>
#include<string>
#include <math.h>
#include <algorithm>
using namespace std;

void swap(int* a, int* b){
    int t = *a;
    *a = *b;
    *b = t;
}

void printArray(int arr[], int n){
    int i;
    cout << "[ ";
    for (i = 0; i < n; i++){
        cout << arr[i] << " ";
    }
    cout <<"]"<< endl;
}
 
int partition (int arr[], int low, int high){
    int pivot = arr[high];
    int i = (low - 1);
 
    for (int j = low; j <= high - 1; j++){
        if (arr[j] < pivot){
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}
 
void quickSort(int arr[], int low, int high){
    if (low < high){
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int main(){
    string arrN[]={"test1.txt","test2.txt", "test3.txt", "test4.txt","test5.txt", "test6.txt", "test7.txt", "test8.txt", "test9.txt", "test10.txt", "test11.txt", "test12.txt", "test13.txt", "test14.txt","test15.txt"};
	int arr2[]={100,1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 40000,50000};
    for(int j=0;j<15;j++){
        double ltiempos[5];
        for(int m=0;m<5;m++){
            unsigned t0, t1;
            int n=arr2[j];
            int arr[n];
            
            ifstream archivo(arrN[j]);
            string texto;

            int i=0;
            while(getline(archivo,texto)){
                arr[i]=stoi(texto);
                i=i+1;
            }
            
            archivo.close();
            //if(m==4){
                //printArray(arr,n);
            //}
            t0 = clock();
            quickSort(arr, 0, n - 1);
            t1 = clock();
            double time = (double(t1-t0)/CLOCKS_PER_SEC);
            ltiempos[m]=time;
            //if(m==4){
                //printArray(arr,n);
            //}
        }
        double promf = 0;
        for(int k=0;k<5;k++){
            promf += ltiempos[k];
        }
        promf = promf / 5;
        double dvs = 0;
        for(int p=0;p<5;p++){
            dvs += pow((ltiempos[p]-promf),2);
        }
        dvs=sqrt(dvs/4);
        cout<<"[Array Tamaño:"<<arr2[j]<<"][Promedio de Tiempo de ejecucion:"<<promf<<"][Desviacion Estandar:"<<dvs<<"]"<<endl;
    }	
	return 0;
}

File: Practica 16/Quicksort.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
	"strconv"
	"time"
)

func partition(arr []int, low, high int) ([]int, int) {
	pivot := arr[high]
	i := low
	for j := low; j < high; j++ {
		if arr[j] < pivot {
			arr[i], arr[j] = arr[j], arr[i]
			i++
		}
	}
	arr[i], arr[high] = arr[high], arr[i]
	return arr, i
}

func quickSort(arr []int, low, high int) []int {
	if low < high {
		var p int
		arr, p = partition(arr, low, high)
		arr = quickSort(arr, low, p-1)
		arr = quickSort(arr, p+1, high)
	}
	return arr
}

func printArray(arr []int) {
	fmt.Print("[")
	for i := 0; i < len(arr); i++ {
		fmt.Print(arr[i], " ")
	}
	fmt.Println("]")
}

func main() {
	nombres := []string{"test1.txt", "test2.txt", "test3.txt", "test4.txt", "test5.txt", "test6.txt", "test7.txt", "test8.txt", "test9.txt", "test10.txt", "test11.txt", "test12.txt", "test13.txt", "test14.txt", "test15.txt"}
	tamaños := []int{100, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 40000, 50000}
	// open the file
	for i := 0; i < 15; i++ {
		var ltiempos []float64
		for j := 0; j < 5; j++ {
			var arr []int
			file, err := os.Open(nombres[i])

			//handle errors while opening
			if err != nil {
				log.Fatalf("Error when opening file: %s", err)
			}

			fileScanner := bufio.NewScanner(file)
			for fileScanner.Scan() {
				sv := fileScanner.Text()
				if sv, err := strconv.Atoi(sv); err == nil {
					arr = append(arr, sv)
				}
			}
			file.Close()

			start := time.Now()
			quickSort(arr, 0, len(arr)-1)
			duration := time.Since(start).Seconds()
			ltiempos = append(ltiempos, float64(duration))

		}
		var promf float64 = 0
		for k := 0; k < 5; k++ {
			promf += ltiempos[k]
		}
		promf = promf / 5
		var dvs float64 = 0
		for p := 0; p < 5; p++ {
			dvs += math.Pow(ltiempos[p]-promf, 2)
		}
		dvs = math.Sqrt(dvs / 4)
		fmt.Println("[Array Tamaño:", tamaños[i], "][Promedio de Tiempo de ejecucion:", promf, "][Desviacion Estandar:", dvs, "]")

	}
}
